sqld/src/hrana/http/stream.rs

```diff
@@ -0,0 +1,402 @@
+            Some(encode_baton(
+                self.server,
+                stream.stream_id,
+                stream.baton_seq,
+            ))
+        } else {
+            None
+        }
+    }
+}
+
+impl<'srv> Drop for Guard<'srv> {
+    fn drop(&mut self) {
+        let stream = self.stream.take().unwrap();
+        let stream_id = stream.stream_id;
+
+        let mut state = self.server.stream_state.lock();
+        let Some(handle) = state.handles.remove(&stream_id) else {
+            panic!("Dropped a Guard for stream {stream_id}, \
+                but Server does not contain a handle to it");
+        };
+        if !matches!(handle, Handle::Acquired) {
+            panic!(
+                "Dropped a Guard for stream {stream_id}, \
+                but Server contained handle that is not acquired"
+            );
+        }
+
+        if self.release {
+            state.handles.insert(stream_id, Handle::Available(stream));
+            mark_expire(&mut state, stream_id);
+            tracing::debug!("Stream {stream_id} was released for further use");
+        } else {
+            tracing::debug!("Stream {stream_id} was closed");
+        }
+    }
+}
+
+fn gen_stream_id(state: &mut ServerStreamState) -> u64 {
+    for _ in 0..10 {
+        let stream_id = rand::random();
+        if !state.handles.contains_key(&stream_id) {
+            return stream_id;
+        }
+    }
+    panic!("Failed to generate a free stream id with rejection sampling")
+}
+
+/// Encodes the baton.
+///
+/// The baton is base64-encoded byte string that is composed from:
+///
+/// - payload (16 bytes):
+///     - `stream_id` (8 bytes, big endian)
+///     - `baton_seq` (8 bytes, big endian)
+/// - MAC (32 bytes): an authentication code generated with HMAC-SHA256
+///
+/// The MAC is used to cryptographically verify that the baton was generated by this server. It is
+/// unlikely that we ever issue the same baton twice, because there are 2^128 possible combinations
+/// for payload (note that both `stream_id` and the initial `baton_seq` are generated randomly).
+fn encode_baton(server: &Server, stream_id: u64, baton_seq: u64) -> String {
+    let mut payload = [0; 16];
+    payload[0..8].copy_from_slice(&stream_id.to_be_bytes());
+    payload[8..16].copy_from_slice(&baton_seq.to_be_bytes());
+
+    let mut hmac = hmac::Hmac::<sha2::Sha256>::new_from_slice(&server.baton_key).unwrap();
+    hmac.update(&payload);
+    let mac = hmac.finalize().into_bytes();
+
+    let mut baton_data = [0; 48];
+    baton_data[0..16].copy_from_slice(&payload);
+    baton_data[16..48].copy_from_slice(&mac);
+    BASE64_STANDARD_NO_PAD.encode(baton_data)
+}
+
+/// Decodes a baton encoded with `encode_baton()` and returns `(stream_id, baton_seq)`. Always
+/// returns a [`ProtocolError::BatonInvalid`] if the baton is invalid, but it attaches an anyhow
+/// context that describes the precise cause.
+fn decode_baton(server: &Server, baton_str: &str) -> Result<(u64, u64)> {
```

Comment:
These are plain functions, what makes them look like methods that pretend to be methods? In general, I prefer to use plain function, unless there is a good reason to use a method.

For example, with the `Guard` struct it is very convenient to call functions like `let db = stream_guard.get_db()?` (rather than `let db = stream::guard_get_db(stream_guard)?`), but with functions like `gen_stream_id()`, `decode_baton()` and `pump_expire()`, I see no need to write them as `impl`-s on their first argument; it would only add syntactical noise (the `impl` and the extra indentation) and make it a teeny bit harder to follow.

In fact, that's one thing that I really like about Rust: it allows me to write code as modules with functions (and use methods when it's convenient), but it does not force me to implement everything as a method, like Java.